phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/controllers/config/CloseChannelsConfigurationController.kt

```diff
@@ -46,34 +46,36 @@ class AppCloseChannelsConfigurationController(
         else -> null
     }
 
-    fun isMutualClosable(channelInfoStatus: ChannelInfoStatus): Boolean = when (channelInfoStatus) {
+    private fun isMutualClosable(channelInfoStatus: ChannelInfoStatus): Boolean = when (channelInfoStatus) {
         ChannelInfoStatus.Normal -> true
         else -> false
     }
 
-    fun isForceClosable(channelInfoStatus: ChannelInfoStatus): Boolean = when (channelInfoStatus) {
+    private fun isForceClosable(channelInfoStatus: ChannelInfoStatus): Boolean = when (channelInfoStatus) {
         ChannelInfoStatus.Normal -> true
         ChannelInfoStatus.Offline -> true
         ChannelInfoStatus.Syncing -> true
         else -> false
     }
 
-    fun isClosable(channelInfoStatus: ChannelInfoStatus): Boolean = if (isForceClose) {
+    private fun isClosable(channelInfoStatus: ChannelInfoStatus): Boolean = if (isForceClose) {
         isForceClosable(channelInfoStatus)
     } else {
         isMutualClosable(channelInfoStatus)
     }
 
-    fun isClosable(channel: ChannelState): Boolean = channelInfoStatus(channel)?.let {
+    private fun isClosable(channel: ChannelState): Boolean = channelInfoStatus(channel)?.let {
         isClosable(it)
     } ?: false
 
     init {
         launch {
             peerManager.getPeer().channelsFlow.collect { channels ->
 
+                val closingChannelIdsCopy = closingChannelIds?.toSet()
+
                 val updatedChannelsList = channels.filter {
-                    closingChannelIds?.let { set ->
+                    closingChannelIdsCopy?.let { set ->
```

Comment:
This can be simplified to :

```kt
closingChannelIdsCopy?.contains(it.key) ?: true
```

By the way, the `private var closingChannelIds` variable in the controller that is updated/consumed in different coroutines looks a bit fishy. There should be a `FIXME` comment so that it can be fixed in a later PR.

Reply:
I've never been happy with the `CloseChannelsConfigurationController`. The whole thing just feels brittle to me.

From a UI perspective, the user is asking to drain his/her wallet. So they would expect that:

- we show them the total balance (in sats) they should receive
- we close every single channel, and broadcast the appropriate transactions
- any deviation from this means an error/warning screen is displayed

But the code is more complicated. First of all, `Peer` requires that we send a COMMAND for each channel that we want to close. But we don't get a response from this command - we have to listen until we see the channel's state transition to CLOSED.

(This is why we were tracking which channels received the command. So the UI could tell when the requested channels had transitioned.)

But the whole process starts with this brittle-looking piece of code:

```kotlin
val channelsToClose = peer.channels.filter {
    when (it) {
      is Normal -> true
      else -> false
    }
}
```

So if any channels had transitioned to Offline, or some other state, then we don't close them. Which makes sense from the point-of-view of the protocol. But not from the point-of-view of the user. If the device goes offline during the process, we shouldn't just skip some channels. And if there are other edge-cases, the user should be alerted about this.

Right now the shared code just kinda does a fire-and-forget of the CMD_CLOSE. But I think there should be a lot more logic here.

Reply:
Fixed in a945b4d1